src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/FieldOnTypeBuilderInstantiation.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Globalization;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class FieldOnTypeBuilderInstantiation : FieldInfo
+    {
+        #region Private Static Members
+        internal static FieldInfo GetField(FieldInfo Field, TypeBuilderInstantiation type)
+        {
+            FieldInfo m;
+
+            // This ifdef was introduced when non-generic collections were pulled from
+            // silverlight. See code:Dictionary#DictionaryVersusHashtableThreadSafety
+            // for information about this change.
+            //
+            // There is a pre-existing race condition in this code with the side effect
+            // that the second thread's value clobbers the first in the hashtable. This is
+            // an acceptable race condition since we make no guarantees that this will return the
+            // same object.
+            //
+            // We're not entirely sure if this cache helps any specific scenarios, so
+            // long-term, one could investigate whether it's needed. In any case, this
+            // method isn't expected to be on any critical paths for performance.
+            if (type.m_hashtable.Contains(Field))
+            {
+                m = (type.m_hashtable[Field] as FieldInfo)!;
+            }
+            else
+            {
+                m = new FieldOnTypeBuilderInstantiation(Field, type);
+                type.m_hashtable[Field] = m;
+            }
+
+            return m;
+        }
+        #endregion
+
+        #region Private Data Members
+        private FieldInfo m_field;
```

Comment:
Nit: You can rename these to `_field`, etc. while you are on it to make it follow repo naming convention.